src/libstore/store-api.cc

```diff
@@ -686,38 +686,56 @@ static bool goodStorePath(const StorePath & expected, const StorePath & actual)
 
-    try {
-        {
-            auto res = state.lock()->pathInfoCache.get(std::string(storePath.to_string()));
-            if (res && res->isKnownNow()) {
-                stats.narInfoReadAverted++;
-                if (!res->didExist())
-                    throw InvalidPath("path '%s' is not valid", printStorePath(storePath));
-                return callback(ref<const ValidPathInfo>(res->value));
-            }
+    {
+        auto res = state.lock()->pathInfoCache.get(std::string(storePath.to_string()));
+        if (res && res->isKnownNow()) {
+            stats.narInfoReadAverted++;
+            if (res->didExist())
+                return std::make_optional(res->value);
+            else
+                return std::make_optional(nullptr);
         }
+    }
 
-        if (diskCache) {
-            auto res = diskCache->lookupNarInfo(getUri(), hashPart);
-            if (res.first != NarInfoDiskCache::oUnknown) {
-                stats.narInfoReadAverted++;
-                {
-                    auto state_(state.lock());
-                    state_->pathInfoCache.upsert(std::string(storePath.to_string()),
-                        res.first == NarInfoDiskCache::oInvalid ? PathInfoCacheValue{} : PathInfoCacheValue{ .value = res.second });
-                    if (res.first == NarInfoDiskCache::oInvalid ||
-                        !goodStorePath(storePath, res.second->path))
-                        throw InvalidPath("path '%s' is not valid", printStorePath(storePath));
-                }
-                return callback(ref<const ValidPathInfo>(res.second));
+    if (diskCache) {
+        auto res = diskCache->lookupNarInfo(getUri(), hashPart);
+        if (res.first != NarInfoDiskCache::oUnknown) {
+            stats.narInfoReadAverted++;
+            {
+                auto state_(state.lock());
+                state_->pathInfoCache.upsert(std::string(storePath.to_string()),
+                    res.first == NarInfoDiskCache::oInvalid ? PathInfoCacheValue{} : PathInfoCacheValue{ .value = res.second });
+                if (res.first == NarInfoDiskCache::oInvalid ||
+                    !goodStorePath(storePath, res.second->path))
+                    return std::make_optional(nullptr);
             }
+            assert(res.second);
+            return std::make_optional(res.second);
         }
+    }
+
+    return std::nullopt;
+}
+
 
+void Store::queryPathInfo(const StorePath & storePath,
+    Callback<ref<const ValidPathInfo>> callback) noexcept
+{
+    auto hashPart = std::string(storePath.hashPart());
+
+    try {
+        auto r = queryPathInfoFromClientCache(storePath);
+        if (r.has_value()) {
+            std::shared_ptr<const ValidPathInfo> & info = *r;
+            if (info)
+                return callback(ref(info));
+            else
+                throw InvalidPath("path '%s' is not valid", printStorePath(storePath));
+        }
     } catch (...) { return callback.rethrow(); }
```

Comment:
(nit) do we still need the try-catch? Can't we have something like below instead?

```suggestion
{
    auto hashPart = std::string(storePath.hashPart());

    auto r = queryPathInfoFromClientCache(storePath);
    if (r.has_value()) {
        std::shared_ptr<const ValidPathInfo> & info = *r;
        if (info)
            return callback(ref(info));
        else
            return callback.rethrow(std::make_exception_ptr<InvalidPath>("path '%s' is not valid", printStorePath(storePath)));
    }
```

Reply:
It does also catch other exceptions, which is important to the contract.
Although I've removed the unchecked `noexcept` modifier (I don't think exceptions should be "upgraded" to crashes), it's still expected for Callback based functions not to throw.